apps/shared/avifpng.c

```diff
@@ -17,6 +20,119 @@ typedef png_bytep png_iccp_datap;
 typedef png_charp png_iccp_datap;
 #endif
 
+// Converts a hexadecimal string which contains 2-byte character representations of hexadecimal values to raw data.
+// hexString may contain values consisting of [A-F][a-f][0-9] in pairs, e.g., 7af2..., separated by any number of newlines.
+// On success the bytes are filled and AVIF_TRUE is returned.
+// AVIF_FALSE is returned if fewer than expectedLength hexadecimal pairs are converted.
+static avifBool avifHexStringToBytes(const char * hexString, size_t expectedLength, avifRWData * bytes)
+{
+    avifRWDataRealloc(bytes, expectedLength);
+    const char * src = hexString;
+    size_t actualLength = 0;
+    for (; (actualLength < expectedLength) && (*src != '\0'); ++actualLength) {
+        while (*src == '\n') {
+            ++src;
+        }
+        if (!isxdigit(src[0]) || !isxdigit(src[1])) {
+            avifRWDataFree(bytes);
+            fprintf(stderr, "Exif extraction failed: invalid token at " AVIF_FMT_ZU "\n", actualLength);
+            return AVIF_FALSE;
+        }
+        const char twoHexDigits[] = { src[0], src[1], '\0' };
+        bytes->data[actualLength] = (uint8_t)strtol(twoHexDigits, NULL, 16);
+        src += 2;
+    }
+
+    if (actualLength != expectedLength) {
+        avifRWDataFree(bytes);
+        fprintf(stderr, "Exif extraction failed: expected " AVIF_FMT_ZU " tokens but got " AVIF_FMT_ZU "\n", expectedLength, actualLength);
+        return AVIF_FALSE;
+    }
+    return AVIF_TRUE;
+}
+
+// Parses the raw profile string of profileLen characters and extracts the payload.
+static avifBool avifCopyRawProfile(const char * profile, size_t profileLength, avifRWData * payload)
+{
+    // ImageMagick formats 'raw profiles' as "\n<name>\n<length>(%8lu)\n<hex payload>\n".
+    if (!profile || (profileLength < 3) || (profile[0] != '\n')) {
+        fprintf(stderr, "Exif extraction failed: truncated or malformed raw profile\n");
+        return AVIF_FALSE;
+    }
+
+    const char * lengthStart = NULL;
+    for (size_t i = 1; i < profileLength; ++i) {
+        if (profile[i] == '\0') {
+            fprintf(stderr, "Exif extraction failed: malformed raw profile, unexpected null-terminating character at " AVIF_FMT_ZU "\n", i);
+            return AVIF_FALSE;
+        } else if (profile[i] == '\n') {
+            if (!lengthStart) {
+                // Skip the name and store the beginning of the string containing the length of the payload.
+                lengthStart = &profile[i + 1];
+            } else {
+                const char * hexPayloadStart = &profile[i + 1];
+                const size_t hexPayloadMaxLength = profileLength - (i + 1);
```

Comment:
`hexPayloadMaxLength` is the maximum length of the hexadecimal string. In line 83, we compare it with `expectedLength`, which is the expected number of converted bytes. These two values differ by a factor of 2, because two hexadecimal digits are converted to one byte.

It seems that in line 83 we should compare `(unsigned long)expectedLength > (hexPayloadMaxLength / 2)` instead.